src/Microsoft.PowerApps.TestEngine/TestInfra/TestRecorder.cs

```diff
@@ -0,0 +1,1062 @@
+                    const r = Math.random() * 16 | 0, v = c === 'x' ? r : (r & 0x3 | 0x8);
+                    return v.toString(16);
+            }});
+
+            fetch('https://{0}/testengine/audio/start', {{
+                method: 'POST',
+                body: JSON.stringify({{ 'startDateTime': new Date().toISOString(), 'audioSessionId': document.TestEngineAudioSessionId  }}),
+                headers: {{
+                    'Content-Type': 'application/json'
+                }}
+            }});
+
+            // When recording stops, create an audio file
+            mediaRecorder.addEventListener('stop', () => {{
+                const audioBlob = new Blob(audioChunks, {{ type: 'audio/wav' }});
+                const audioUrl = URL.createObjectURL(audioBlob);
+                audioPlayback.src = audioUrl;
+
+                // Post the recorded audio to an API
+                fetch('https://{0}/testengine/audio/upload', {{
+                    method: 'POST',
+                    body: audioBlob,
+                    headers: {{
+                        'Content-Type': 'audio/wav',
+                        'endDateTime': new Date().toISOString(),
+                        'audioSessionId': document.TestEngineAudioSessionId
+                    }}
+                }}).then(response => {{
+                    if (response.ok) {{
+                        feedback.textContent = 'Audio uploaded successfully!';
+                    }} else {{
+                        feedback.textContent = 'Failed to upload audio.';
+                    }}
+                }}).catch(error => {{
+                    feedback.textContent = 'Error uploading audio: ' + error;
+                }});
+            }});
+        }});
+
+        // Function to stop recording
+        stopButton.addEventListener('click', () => {{
+            mediaRecorder.stop();
+            startButton.disabled = false;
+            stopButton.disabled = true;
+            feedback.textContent = 'Recording stopped. Uploading audio...';
+        }});
+
+        // Function to close the dialog
+        closeButton.addEventListener('click', () => {{
+            document.body.removeChild(dialog);
+        }});
+    }}
+}});
+";
+
+            await _infra.Page.EvaluateAsync(string.Format(recordingJavaScript, feedbackHost));
+
+            // Add recording if page is reloaded
+            _infra.Page.Load += async (object sender, IPage e) =>
+            {
+                await _infra.Page.EvaluateAsync(string.Format(recordingJavaScript, feedbackHost));
+            };
+        }
+
+        ///<summary>
+        /// Sets up the TestRecorder by subscribing to page mouse events.
+        ///</summary>
+        public void SetupMouseMonitoring()
+        {
+            var page = _infra.Page;
+
+            var feedbackUrl = new Uri(new Uri($"https://{new Uri(_testState.GetDomain()).Host}"), new Uri("testengine", UriKind.Relative));
```

Comment:
should this be extracted out to be common for all functions

Reply:
Initial focus will be for MDA provider. Will need to create new feature requests to test and refactor across multiple providers